utils/phpstan-extensions/src/Rule/CheckGetNodeTypesReturnPhpParserNodeRule.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\PHPStanExtensions\Rule;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\Array_;
+use PhpParser\Node\Name\FullyQualified;
+use PhpParser\Node\Stmt\ClassMethod;
+use PhpParser\Node\Stmt\Return_;
+use PhpParser\NodeFinder;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use PHPStan\ShouldNotHappenException;
+
+/**
+ * @see \Rector\PHPStanExtensions\Tests\Rule\CheckGetNodeTypesReturnPhpParserNodeRule\CheckGetNodeTypesReturnPhpParserNodeRuleTest
+ */
+final class CheckGetNodeTypesReturnPhpParserNodeRule implements Rule
+{
+    /**
+     * @var string
+     */
+    public const ERROR = "%s::getNodeTypes() returns class names that are not instances of %s:\n%s";
+
+    public function getNodeType(): string
+    {
+        return ClassMethod::class;
+    }
+
+    /**
+     * @param ClassMethod $node
+     * @return string[]
+     */
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if ($node->name->toString() !== 'getNodeTypes') {
+            return [];
+        }
+
+        if ($this->shouldSkip($node, $scope)) {
+            return [];
+        }
+
+        $incorrectClassNames = $this->getIncorrectClassNames($node);
+        if ($incorrectClassNames === null) {
+            return [];
+        }
+
+        return [sprintf(
+            self::ERROR,
+            $scope->getClassReflection()->getName(),
+            Node::class,
+            implode(",\n", $incorrectClassNames)
+        )];
+    }
+
+    private function shouldSkip(ClassMethod $classMethod, Scope $scope): bool
+    {
+        if ($scope->getClassReflection() === null) {
+            return true;
+        }
+
+        if ($scope->getClassReflection()->isInterface()) {
+            return true;
+        }
+
+        $nodeFinder = new NodeFinder();
+        /** @var Return_[] $returns */
+        $returns = $nodeFinder->findInstanceOf($classMethod, Return_::class);
+
+        if (count($returns) === 1) {
+            return false;
+        }
+
+        throw new ShouldNotHappenException('More than one return statement exist');
```

Comment:
@TomasVotruba, at the moment, we don't have such cases. Do you think I have to add handling of such a situation into this rule or some separate rule?

Reply:
I think exception here is the best solution :+1: